File: solutions/0042-trapping-rain-water/solution.java
class Solution {
    public int trap(int[] height) 
    {
        if (height.length == 0) return 0;

        int l = 0, r = height.length - 1;

        int maxLeft = height[l], maxRight = height[r];

        int result = 0;

        while (l < r)
        {
            if (maxLeft < maxRight)
            {
                l++;
                maxLeft = Math.max(height[l], maxLeft);
                result += maxLeft - height[l];
            } else
            {
                r--;
                maxRight = Math.max(height[r], maxRight);
                result += maxRight - height[r];
            }
        }    

        return result;
    }
}

File: solutions/0042-trapping-rain-water/solution.cpp
class Solution {
public:
    int trap(vector<int>& height) {
        int l = 0;
        int r = height.size() - 1;
        
        int max_l = height[l];
        int max_r = height[r];

        int res = 0;

        while (l < r) {
            if (max_l < max_r) {
                l++;
                max_l = max(max_l, height[l]);
                res += max_l - height[l];
            } else {
                r--;
                max_r = max(max_r, height[r]);
                res += max_r - height[r]; 
            }
        }

        return res;
    }
};
